articles/WideStrings.md

```diff
@@ -0,0 +1,186 @@
+* [http://stackoverflow.com/questions/4588302/why-isnt-wchar-t-widely-used-in-code-for-linux-related-platforms](http://stackoverflow.com/questions/4588302/why-isnt-wchar-t-widely-used-in-code-for-linux-related-platforms)
+* [http://www.diveintopython3.net/strings.html](http://www.diveintopython3.net/strings.html)
+* [http://stackoverflow.com/questions/402283/stdwstring-vs-stdstring](http://stackoverflow.com/questions/402283/stdwstring-vs-stdstring)
+* [https://utf8everywhere.org/](http://utf8everywhere.org/)
+
+
+## Unicode Characters in Strings
+Two goals for ROS 2 strings are to be compatible with ROS 1 strings, and compatible with the DDS wire format.
+ROS 1 says string fields are to contain ASCII encoded data, but allows UTF-8.
+DDS-XTYPES mandates UTF-8 be used as the encoding of IDL type `string`.
+To be compatibile with both, in ROS 2 the content of a `string` is expected to be UTF-8.
+
+## Wide Strings
+ROS 2 messages will have a new [primitive field type](/articles/interface_definition.html) `wstring`.
+This purpose is to allow ROS 2 nodes to comminicate with non-ROS DDS entities using an IDL containing a `wstring` field.
+The encoding of data in this type should be UTF-16 to match DDS-XTYPES 1.2.
+Since both UTF-8 and UTF-16 can encode the same code points, new ROS 2 messages should prefer `string` over `wstring`.
+
+## Encodings are Required but not Guaranteed to be Enforced
+`string` and `wstring` are required to be UTF-8 and UTF-16, but the requirement may not be enforced.
+Since ROS 2 is targeting resource constrained systems, it is left to the rmw implementation to choose whether to enforce the encoding.
+Further, since many users will write code to check that a string contains valid data, checking again in lower layers may not be necessary in some cases.
+
+If a `string` or `wstring` field is populated with the wrong encoding then the behavior is undefined.
+It is possible the rmw implementation may allow invalid strings to be passed through to subscribers.
+Each subscriber is responsible for detecting invalid strings and deciding how to handle them.
+For example, subscribers like `ros2 topic echo` may echo the bytes in hexadecimal.
+
+The IDL specification forbids `string` from containing `NULL` values.
+To be compatible, a ROS message `string` field must not contain zero bytes, and a `wstring` field must not contain zero words.
+This restriction will be enforced.
+
+## Unicode Strings Across ROS 1 Bridge
+Since ROS 1 and 2 both allow `string` to be UTF-8, the ROS 1 bridge will pass values unmodified between them.
+If a message with a string field fails to serialize because the content is not legal UTF-8 then the default behavior will be to drop the entire message.
+Other strategies like replacing invalid bytes could unintentionally change the meaning, so they will be opt-in if available at all.
+
+If a ROS 2 message has a field of type `wstring` then the bridge will attempt to convert it from UTF-16 to UTF-8.
```

Comment:
Is that currently implemented?

Reply:
Probably not. Are you asking for a note to be added saying it's not implemented yet?

Reply:
Yes, e.g.

Reply:
Added note in 9818383

Reply:
I don't think anything needs to happen for `strings`. They are just copied one-to-one already. So the note only applies to `wstrings` afaik.

Reply:
Limited note to `wstring` fields in ecf689b